docs/reference/scripting/apis/create-stored-script-api.asciidoc

```diff
@@ -0,0 +1,114 @@
+[[create-stored-script-api]]
+=== Create or update stored script API
+++++
+<titleabbrev>Create or update stored script</titleabbrev>
+++++
+
+Creates or updates a <<script-stored-scripts,stored script>> or
+stored <<search-template,search template>>.
+
+[source,console]
+----
+PUT _scripts/my-stored-script
+{
+  "script": {
+    "lang": "painless",
+    "source": """
+      TimestampHour date =  doc['@timestamp'].value; 
+      return date.getHour()
+    """
+  }
+}
+----
+
+[[create-stored-script-api-request]]
+==== {api-request-title}
+
+`PUT _search/<script-id>`
+
+`POST _scripts/<script-id>`
+
+`PUT _search/<script-id>/<context>`
+
+`POST _scripts/<script-id>/<context>`
+
+[[create-stored-script-api-prereqs]]
+==== {api-prereq-title}
+
+* If the {es} {security-features} are enabled, you must have the `manage`
+<<privileges-list-cluster,cluster privilege>> to use this API.
+
+[[create-stored-script-api-path-params]]
+==== {api-path-parms-title}
+
+`<script-id>`::
+(Required, string)
+Name or identifier for the stored script or search template. Must be unique
+within the cluster.
+
+`<context>`::
+(Optional, string)
+{painless}/painless-contexts.html[Painless context] in which the script runs.
+You can only specify this parameter for Painless scripts.
```

Comment:
Contexts can be applied to any scripting language. It's just that expressions and mustache have a limited subset of contexts they support. As a long term change we should move painless-contexts into script-contexts. (I have spoken with @lockewritesdocs about this recently). This context parameter is used to immediately try to compile a script when it's stored which guarantees that it won't have compile-time errors when it's used for that specific context.

Reply:
Thanks for the explanation @jdconrad. That makes sense. I pushed some commits to fix the wording here and do some light reorg for future planned API docs. We can xref to the context docs here once those are finalized.